man/man1/weyes.1
.TH WEYES 1 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
weyes \- eyes that follow mouse (xeyes for W)
.SH SYNOPSIS
weyes [-h] [-r] [-g \fIgeometry\fP]
.SH DESCRIPTION
.I Weyes
opens a window with two eyes which will always look at the mouse, so if you
should once have lost it in the pixel chaos on your screen... :)
.PP
Weyes supports the \fIgeometry\fP command line option, for details on
how to use that take a look at the appropriate manpage.
.PP
Use the \fI-r\fP option to put \fIweyes\fP onto root window (my favorite).
This way W server doesn't need to allocate window bitmap for it.
.SH SEE ALSO
scan_geometry(3)
.SH LAST CHANGED
30th. Nov. 1997 by Eero Tamminen


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


apps/weyes.c
/*
 * weyes.c, a part of the W Window System
 *
 * Copyright (C) 1994-1997 by Torsten Scherer and
 * Copyright (C) 1997-1998 by Eero Tamminen
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- show eyes that follow the mouse pointer
 *
 * CHANGES
 * 11/97, ++eero:
 * - resizable
 * - may be quit with 'q' key
 * - uses ellipses instead of circles
 * - eye highlights
 * - may be put on root
 * 12/97, jps:
 * - adapted to the swallow feature of wlaunch
 * - added long options
 * 2/98, ++eero:
 * - replaced trigonometrics with integer square root
 */

#include <stdio.h>
#include <stdlib.h>
#include <Wlib.h>
#include <getopt.h>


static struct option long_options [] = {
  { "geometry", required_argument, NULL, 'g' },
  { "help", no_argument, NULL, 'h' },
  { "parent", required_argument, NULL, 'p'},
  { "root", no_argument, NULL, 'r' },
  { NULL, 0, NULL, 0}
};

#define MIN_XSIZE 48
#define MIN_YSIZE 48

typedef struct {
  short midx, midy;
  short pupx, pupy;
} EYE;


static WWIN *win;
static short rootw, rooth, winw, winh, rx, ry, rxpup, rypup;
static EYE lefteye, righteye;
static int use_root;


static void updateeye(EYE *eye)
{
  short dx, dy, scale;

  w_querymousepos(win, &dx, &dy);
  dx -= eye->midx;
  dy -= eye->midy;

  scale = isqrt((ulong)dx*dx + (ulong)dy*dy);
  if (scale > rx) {
    dx = eye->midx + (short)((long)dx * rx / scale);
  } else {
    dx += eye->midx;
  }
  if (scale > ry) {
    dy = eye->midy + (short)((long)dy * ry / scale);
  } else {
    dy += eye->midy;
  }

  /* move the pupille */
  if ((dx != eye->pupx) || (dy != eye->pupy)) {
    w_pbox(win, eye->pupx - rxpup/2, eye->pupy - rypup/2, rxpup/3, rypup/3);
    w_pellipse(win, eye->pupx, eye->pupy, rxpup, rypup);
    eye->pupx = dx;
    eye->pupy = dy;
    w_pellipse(win, dx, dy, rxpup, rypup);
    w_pbox(win, dx - rxpup/2, dy - rypup/2, rxpup/3, rypup/3);
  }
}

static void draw_eyes(int xp, int yp)
{
  if (!use_root) {
    w_setmode(win, M_DRAW);
    w_pbox(win, 0, 0, winw, winh);
  }

  rx = winw/4;
  ry = winh/2;
  lefteye.midx = lefteye.pupx = xp + rx;
  lefteye.midy = lefteye.pupy = yp + ry;
  righteye.midx = righteye.pupx = xp + 3*rx;
  righteye.midy = righteye.pupy = yp + ry;

  rx--;
  ry--;
  w_setmode(win, M_CLEAR);
  w_pellipse(win, lefteye.midx, lefteye.midy, rx, ry);
  w_pellipse(win, righteye.midx, righteye.midy, rx, ry);

  if (use_root) {
    w_setmode(win, M_DRAW);
    w_ellipse(win, lefteye.midx, lefteye.midy, rx, ry);
    w_ellipse(win, righteye.midx, righteye.midy, rx, ry);
  }

  rxpup = rx / 4;
  rypup = ry / 4;
  rx -= (rxpup + 2);
  ry -= (rypup + 2);

  w_setmode(win, M_INVERS);
  w_pellipse(win, lefteye.pupx, lefteye.pupy, rxpup, rypup);
  w_pellipse(win, righteye.pupx, righteye.pupy, rxpup, rypup);
  w_pbox(win, lefteye.pupx - rxpup/2, lefteye.pupy - rypup/2, rxpup/3, rypup/3);
  w_pbox(win, righteye.pupx - rxpup/2, righteye.pupy - rypup/2, rxpup/3, rypup/3);
}

static void check_size(short *winw, short *winh)
{
  if (*winw < MIN_XSIZE) {
    *winw = MIN_XSIZE;
  }
  if (*winh < MIN_YSIZE) {
    *winh = MIN_YSIZE;
  }
}



static void usage (char *name) {
  fprintf (stderr, "\
Usage: %s [OPTION]...\n\
A well-known program for a well-known purpose.\n\
\n\
  -g, --geometry             set size and position\n\
  -h, --help                 display this help and exit\n\
  -r, --root                 use root window for eyes\n\
\n\
", name);
}



int main(int argc, char **argv)
{
  short ewinw = MIN_XSIZE;
  short ewinh = MIN_YSIZE;
  short xp = UNDEF;
  short yp = UNDEF;
  short show_usage = 0;
  ulong parentid = 0;
  WWIN *pwin = NULL;
  WSERVER *wserver;
  WEVENT *ev;
  int c;

  while ((c = getopt_long (argc, argv, "g:hrp:", long_options, NULL)) != -1) {
    switch (c) {

      case 'g':
        scan_geometry (optarg, &winw, &winh, &xp, &yp);
        break;

      case 'p':
        sscanf (optarg, "%lx", &parentid);
        break;

      case 'r':
	use_root = 1;
	break;

      default:
        show_usage = 1;
	break;
    }
  }

  if (show_usage) {
    usage (argv[0]);
    return -1;
  }

  if (!(wserver = w_init())) {
    fprintf(stderr, "error: %s can't connect to wserver\n", argv[0]);
    return -1;
  }
  rootw = wserver->width;
  rooth = wserver->height;

  if (parentid) {
    /* have to be connected to W server first */
    pwin = w_winFromID(parentid);
    winw = pwin->width;
    winh = pwin->height;
  }

  check_size(&winw, &winh);

  if (use_root) {
    win = WROOT;
    ewinw = winw + 2;
    ewinh = winh + 2;
    w_setlinewidth(win, 1);	/* just in case */
  } else {
    if (pwin) {
      if (!(win = w_createChild (pwin, winw, winh, W_NOBORDER))) {
        fprintf(stderr, "error: %s can't create child window\n", argv[0]);
        exit(-1);
      }
    }
    else { 
      if (!(win = w_create(winw, winh, W_MOVE | W_RESIZE | EV_KEYS))) {
        fprintf(stderr, "error: %s can't create window\n", argv[0]);
        exit(-1);
      }
    }
    w_querywinsize(win, 1, &ewinw, &ewinh);
  }

  if (xp < 0 && xp != UNDEF) {
    xp = rootw - ewinw - xp;
  }
  if (yp < 0 && yp != UNDEF) {
    yp = rooth - ewinh - yp;
  }

  if (use_root) {
    if (xp == UNDEF || yp == UNDEF) {
      xp = 0;
      yp = 0;
    }
    draw_eyes(xp, yp);
  } else {
    draw_eyes(0, 0);
    if (pwin) {
      xp = yp = 0;
    }
    if (w_open(win, xp, yp) < 0) {
      fprintf(stderr, "error: %s can't open window\n", argv[0]);
      w_exit();
      return -1;
    }
  }

  while (42) {

    updateeye(&lefteye);
    updateeye(&righteye);

    /* watch out for `exit' event */
    ev = w_queryevent(NULL, NULL, NULL, 200);
    if(!ev)
      continue;

    switch(ev->type)
    {
      case EVENT_GADGET:
      case EVENT_KEY:
        if (ev->key != GADGET_EXIT && ev->key != 'q')
	  break;
	w_exit();
	return 0;

      case EVENT_RESIZE:
	winw = ev->w;
	winh = ev->h;
	check_size(&winw, &winh);
	w_move (ev->win, ev->x, ev->y);
	w_resize (ev->win, winw, winh);

	/* updates global size variables too */
	draw_eyes(0, 0);
	break;
    }
  }
}
